test/functional/p2p_blocksonly.py

```diff
@@ -69,18 +53,58 @@ def run_test(self):
         peer_2_info = self.nodes[0].getpeerinfo()[1]
         assert_equal(peer_2_info['whitelisted'], True)
         assert_equal(peer_2_info['permissions'], ['noban', 'forcerelay', 'relay', 'mempool', 'download'])
-        assert_equal(self.nodes[0].testmempoolaccept([sigtx])[0]['allowed'], True)
-        txid = self.nodes[0].testmempoolaccept([sigtx])[0]['txid']
+        assert_equal(self.nodes[0].testmempoolaccept([tx_hex])[0]['allowed'], True)
 
         self.log.info('Check that the tx from forcerelay first_peer is relayed to others (ie.second_peer)')
         with self.nodes[0].assert_debug_log(["received getdata"]):
-            first_peer.send_message(msg_tx(FromHex(CTransaction(), sigtx)))
+            first_peer.send_message(msg_tx(tx))
             self.log.info('Check that the forcerelay peer is still connected after sending the transaction')
             assert_equal(first_peer.is_connected, True)
             second_peer.wait_for_tx(txid)
             assert_equal(self.nodes[0].getmempoolinfo()['size'], 1)
         self.log.info("Forcerelay peer's transaction is accepted and relayed")
 
+        self.nodes[0].disconnect_p2ps()
+        self.nodes[0].generate(1)
+
+    def blocks_relay_conn_tests(self):
+        self.log.info('Tests with node in normal mode with block-relay-only connections')
+        self.restart_node(0, ["-noblocksonly"])  # disables blocks only mode
+        assert_equal(self.nodes[0].getnetworkinfo()['localrelay'], True)
+
+        # Ensure we disconnect if a block-relay-only connection sends us a transaction
+        self.nodes[0].add_outbound_p2p_connection(P2PInterface(), connection_type="blockrelay")
+        assert_equal(self.nodes[0].getpeerinfo()[0]['relaytxes'], False)
+        _, txid, tx_hex = self.check_p2p_tx_violation(index=2)
+
+        self.log.info("Check that txs from RPC are not sent to blockrelay connection")
+        conn = self.nodes[0].add_outbound_p2p_connection(P2PTxInvStore(), connection_type="blockrelay")
+
+        # bump time forward to ensure nNextInvSend timer pops
+        self.nodes[0].setmocktime(int(time.time()) + 5)
+
+        assert_equal(self.nodes[0].testmempoolaccept([tx_hex])[0]['allowed'], True)
```

Comment:
Why is this needed? If the test fails, is it not ok to fail for `sendrawtransaction` validation/mempool reasons?

Reply:
I don't really understand this question- the test is essentially checking that there is a well formed transaction, submits it to the node, then checks if the p2p connection received the transaction. the goal is to isolate (as much as possible) the reason the conn didn't receive the transaction is due to being a block-relay-only connection.

Reply:
Hm, I guess I found it odd because `sendrawtransaction` will throw a JSONRPCError if validation fails just like `testmempoolaccept` would return False. I don't think they'd have different results unless mempool changes in between calls or something. So it seems like something unnecessary that the test needs to pass.

Reply:
I agree. This doesn't seem necessary. You could just as easily call `sendrawtransaction` and then `getmempoolentry` to confirm that the transaction reached the mempool.

Reply:
ok. I didn't look closely at exactly what errors `testmempoolaccept` returns vs `sendrawtransaction`. I think I probably just copied the existing pattern in the tests. 

I guess I'll just delete this line then? 

^ seems like what @glozow is suggesting. @jnewbery your suggestion just seems like an alternative that's about the same?

Reply:
`sendrawtransaction` error handling is undocumented, but it will indeed return an jsonrpc error when the tx couldn't be added to the mempool

Reply:
Yep, `testmempoolaccept` is unnecessary here. And if you need to make sure it's in mempool then submit + `getmempoolentry` is the move.

Reply:
removed `testmempoolaccept` call